proxy/src/dns.rs

```diff
@@ -60,29 +61,34 @@ impl AsRef<str> for Name {
     }
 }
 
+
 impl Resolver {
     /// Construct a new `Resolver` from the system configuration and Conduit's
     /// environment variables.
     ///
     /// TODO: Make this infallible, like it is in the `domain` crate.
-    pub fn from_system_config_and_env(env_config: &Config) -> Result<Self, ResolveError> {
+    pub fn from_system_config_and_env(env_config: &Config)
+        -> Result<(Self, impl Future<Item = (), Error = ()> + Send), ResolveError> {
         let (config, opts) = trust_dns_resolver::system_conf::read_system_conf()?;
         let opts = env_config.configure_resolver_opts(opts);
         trace!("DNS config: {:?}", &config);
         trace!("DNS opts: {:?}", &opts);
         Ok(Self::new(config, opts))
     }
 
-    pub fn new(config: ResolverConfig,  mut opts: ResolverOpts) -> Self {
+    pub fn new(config: ResolverConfig,  mut opts: ResolverOpts)
+        -> (Self, impl Future<Item = (), Error = ()> + Send)
```

Comment:
Can we give a name to `impl Future<Item = (), Error = ()> + Send`? It's impossible to understand what these functions return by their signatures now.

Reply:
While making these changes in Trust-DNS, I had intended for the background future to be named, but it currently involves a future with a closure that captures a lot of variables. 

I'm fairly sure this could be changed. At the time, however, the code necessary to avoid the closure seemed like it was going to be complex enough that, I opted to return `impl Future<...>` instead. I'd be happy to propose additional changes upstream to rectify this, if you like.

Reply:
I think the current way is OK. I guess there will be named existential types in Rust soon. Until then, we could just add a comment indicating what we'd name the type if we had named existentials available.

Reply:
Support for `impl Trait` in structs could also solve this, as we could wrap it in a `pub struct BackgroundTask(impl Future<...>)`, but I don't think that's likely to be possible, as the struct wouldn't have a known size...

Reply:
OK, let's just add a comment that indicates what we'd call the `impl Future` if we could give it a name and ship this! (I already approved it previously.)

Reply:
Will do. The only reason this hasn't shipped yet is because it's based on #974, which was not previously approved, and I didn't want to merge this as it would also merge those changes.